src/modules/roc_core/target_posix/roc_core/format_time.cpp

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright (c) 2019 Roc authors
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+#include <stdio.h>
+#include <sys/time.h>
+#include <time.h>
+
+#include "roc_core/format_time.h"
+
+namespace roc {
+namespace core {
+
+bool format_time(char* buf, size_t bufsz) {
+    timeval tv;
+    if (gettimeofday(&tv, NULL) == -1) {
+        return false;
+    }
+
+    tm t;
+    if (!localtime_r(&tv.tv_sec, &t)) {
+        return false;
+    }
+
+    size_t off = strftime(buf, bufsz, "%H:%M:%S", &t);
+    if (off == 0) {
+        return false;
+    }
+
+    if (off == bufsz) {
+        return false;
+    }
+
+    if (snprintf(buf + off, bufsz - off, ".%03lu", (unsigned long)tv.tv_usec / 1000)
```

Comment:
What if the provided buffer is too small for the remaining ms?

Reply:
The string will be truncated.

Reply:
This behavior is inconsistent with the function doc, so I fixed it. Now we either write the full string or return an error.

Reply:
Also added some tests.